### PR TITLE
Add photonic ZNE mitigation utilities

### LINKED ISSUE #181: [Feature Request] Design of Zero Noise Extrapolation on photonic devices
This issue is for discussion on the feasibility and related design of Zero Noise Extrapolation (ZNE), a quantum error mitigation (QEM) technique, on photonic devices and hence its adoption in a workflow that includes the use of qlass.

ZNE is routinely applied to the quantum variational eigensolver (VQE) and other variational quantum algorithms (VQA) like the quantum approximate optimization algorithm (QAOA). References can be found also in the [Mitiq Documentation](https://mitiq.readthedocs.io/en/stable/), both in the examples and User guide section ([1](https://mitiq.readthedocs.io/en/stable/examples/vqe-pyquil-demo.html), [2](https://mitiq.readthedocs.io/en/stable/guide/zne-3-options.html)). 

ZNE at its core requires to transform a quantum circuit (or more generally a unitary) into a different but equivalent unitary that takes longer to run on a quantum processing unit (QPU) and thus this scales the noise. ZNE is agnostic with respect to the noise scaling technique and several ones have been proposed, including "digital" ones, that abstract the noise scaling at the level of gates and thus treat the noise scaling as a unitary folding in the quantum circuit, since $U U^\dagger U = U$. Specific instances include local folding (single gates) and global folding (chunks of the quantum circuit). A theoretical summary of folding methods can be found in the Mitiq white paper see [here](https://quantum-journal.org/papers/q-2022-08-11-774/) and in this [other paper on the impact of correlated noise](https://arxiv.org/abs/2201.11792). 

While ZNE is agnostic with respect to the device characteristics, the difficulty in its adoption for photonic QPUs lies in the way quantum circuits are actually implemented in the Knill-Laflamme-Milburn (KLM) scheme in photonics. Measurement and considerable overhead in the runs prevent to apply ZNE to two-qubit gate equivalents (as CNOTs). 

One option could be to apply local folding to single-qubit (single mode) operations, but the issue could be that in integrated photonics circuits the noise impact is too little when adding a few phase shifters.

Alternatively, one could consider transformations at the level of the unitary matrix, $U$, including the possibility to apply the scheme to continuous waveguide arrays as in the 3D architecture. 

Note that in some instances ZNE can be composed with other QEM techniques, such as some readout error mitigation schemes, as shown [here](https://mitiq.readthedocs.io/en/stable/examples/combine_rem_zne.html), which can be relevant to issue #178.

### PR DESCRIPTION
**Pull request by Kitsunp**
Closes #181

### Summary

This PR adds Zero Noise Extrapolation support under `qlass.mitigation`:

- Adds `ZNEMitigator` in `qlass/mitigation/zne.py` for scaled expectation-value execution and zero-noise extrapolation.
- Supports `linear`, `polynomial`, and `exponential` extrapolation.
- Adds `fold_global_interferometer()` for global folding of deterministic Perceval linear optical circuits as `U (U† U)^n`.
- Adds `scale_loss_config()` to scale `HardwareConfig` photon-loss parameters in dB without mutating the original config.
- Exports the new ZNE utilities from `qlass.mitigation`.
- Adds API documentation for `qlass.mitigation`.
- Adds unit tests covering extrapolation, folding, loss scaling, and invalid inputs.

### Verification

```bash
UV_PROJECT_ENVIRONMENT=.venv-wsl uv run --group lint ruff check .
UV_PROJECT_ENVIRONMENT=.venv-wsl uv run --group lint ruff format --check .
UV_PROJECT_ENVIRONMENT=.venv-wsl uv run --group type-check mypy src/qlass
UV_PROJECT_ENVIRONMENT=.venv-wsl uv run --group test pytest --cov -q
```

Result:

```text
99 passed
src/qlass/mitigation/zne.py: 100% coverage
TOTAL coverage: 94%
```

Docs build was also checked:

```bash
UV_PROJECT_ENVIRONMENT=.venv-wsl uv run --group docs -m sphinx -T -b html -d docs/_build/doctrees docs docs/_build/html
```

The build succeeded; existing docutils warnings remain outside this change.

### Notes

The noisy 2-mode ZNE tests use deterministic expectation-value models rather than stochastic shot-based sampling. This keeps the tests stable while verifying the intended ZNE behavior: scaled execution, linear/polynomial extrapolation to the zero-noise intercept, global interferometer folding, and photon-loss configuration scaling.

### AI Transparency

I used OpenAI Codex to help inspect the repository patterns, draft the implementation/tests, and run local verification. I manually reviewed the math, code paths, tests, and command outputs before preparing this PR.

### CODE REVIEW

**Comment by Kitsunp** [comment]
@FarLab Could you take a look at my PR?

**Comment by FarLab** [comment]
@Kitsunp The approach looks like its mathematically correct, but it might be worth adding at least one end-to-end integration test that simulates a real noisy linear optical circuit and verifies that the mitigated energy is physically closer to the ideal energy than the unmitigated one. You can do this by adding a new test.

**Comment by Kitsunp** [comment]
Thanks  @FarLab  , I added an end-to-end Perceval integration test for this.

The new test builds an actual 2-mode linear optical circuit:

```python
circuit = pcvl.Circuit(2).add((0, 1), pcvl.BS.H())
input_state = pcvl.BasicState([1, 0])
```

It then simulates photon loss using `pcvl.Processor("SLOS", ..., noise=pcvl.NoiseModel(transmittance=...))`, with the transmittance computed from a dB loss model:

```python
transmittance = 10 ** (-(loss_db * noise_scale) / 10)
```

The test uses the simple energy observable `E = -<n0>`. The actual values from the test are:

```text
ideal energy:       -0.5000000000000001
unmitigated energy: -0.4830254394949068
mitigated energy:   -0.4999999999999998
```

So the absolute error improves from:

```text
unmitigated error: 0.0169745605050933
mitigated error:   0.0000000000000003
```

The scaled noisy energies used for the extrapolation are:

```text
noise_scale=1.0, transmittance=0.9660508789898133, energy=-0.4830254394949068
noise_scale=2.0, transmittance=0.9332543007969910, energy=-0.4666271503984956
noise_scale=3.0, transmittance=0.9015711376059569, energy=-0.4507855688029785
```

I used exponential extrapolation here because dB-scaled photon loss gives exponential attenuation in the expectation value.

I re-ran the checks locally:

```text
tests/test_zne_mitigation.py: 14 passed
full test suite with coverage: 101 passed
src/qlass/mitigation/zne.py: 100% coverage
TOTAL coverage: 94%
docs build: succeeded, existing warnings remain
```

**Comment by FarLab** [comment]
Thanks for the update @Kitsunp . This is for scaling the transmittance and it seems to work!
Did you try the global folding approach as well? Would be useful to have an end to end test for that.

**Comment by Kitsunp** [comment]
@FarLab Would this end-to-end test cover what you had in mind for the global folding path?

It uses `fold_global_interferometer()` with scaling factors `1, 3, 5`, builds an actual Perceval `Processor("SLOS", ...)`, adds a fixed loss channel after each folded interferometer block, and checks that the ZNE-mitigated energy is closer to the ideal energy than the unmitigated noisy one.

```python
def test_zne_mitigates_global_folded_perceval_interferometer():
    circuit = pcvl.Circuit(2).add((0, 1), pcvl.BS.H())
    input_state = pcvl.BasicState([1, 0])
    component_loss = 0.05
    folded_component_counts = []

    def energy_from_processor(processor):
        processor.with_input(input_state)
        processor.min_detected_photons_filter(0)
        probabilities = processor.probs()["results"]
        return -sum(state[0] * probability for state, probability in probabilities.items())

    def folded_noisy_energy(params, noise_scale):
        folded_circuit = fold_global_interferometer(circuit, noise_scale)
        folded_components = list(folded_circuit)
        folded_component_counts.append(len(folded_components))
        processor = pcvl.Processor("SLOS", 2)

        for modes, component in folded_components:
            processor.add(modes, component)
            for mode in range(2):
                processor.add(mode, pcvl.LC(component_loss))

        return energy_from_processor(processor)

    params = np.array([])
    ideal_energy = energy_from_processor(pcvl.Processor("SLOS", circuit.copy()))
    unmitigated_energy = folded_noisy_energy(params, noise_scale=1.0)
    folded_component_counts.clear()
    mitigator = ZNEMitigator(
        folded_noisy_energy,
        scaling_factors=[1.0, 3.0, 5.0],
        extrapolation_method="exponential",
    )

    mitigated_energy = mitigator.mitigate(params)

    assert folded_component_counts == [1, 3, 5]
    assert abs(mitigated_energy - ideal_energy) < abs(unmitigated_energy - ideal_energy)
    assert np.isclose(mitigated_energy, ideal_energy)
```

The actual values from the test are:

```text
ideal_energy      = -0.5000000000000001
unmitigated       = -0.4750000000000001
mitigated         = -0.5000000000000002

unmitigated error = 0.0250000000000000
mitigated error   = 0.0000000000000001
```

The folded runs used:

```text
scale=1.0 -> folded_components=1, energy=-0.4750000000000001
scale=3.0 -> folded_components=3, energy=-0.4286875000000000
scale=5.0 -> folded_components=5, energy=-0.3868904687499998
```

**Comment by FarLab** [comment]
@Kitsunp Great, that looks good to me!
